File: Demos/Todos/Todos/Fetcher.m
//
//  Fetcher.m
//  Toodoos
//
//  Created by Sasa Prodribaba on 13/12/2018.
//  Copyright © 2018 Sasa Prodribaba. All rights reserved.
//

#import "Fetcher.h"
#import "Serializer.h"

@implementation Fetcher

#define kTodosEndpoint @"https://jsonplaceholder.typicode.com/todos"

- (NSData *)fetchFromURL:(NSURL *)url completion:(fetchCompletion)comp {
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        
        if (!error) {
            comp(data);
        } else {
            NSLog(@"Error: %@", error.localizedDescription);
        }
                                      
    }];
    [dataTask resume];
    
    return 0;
}

- (NSArray *)fetchWithResult:(resultCompletion)comp {
    
    [self fetchFromURL:[NSURL URLWithString:kTodosEndpoint] completion:^(NSData * result) {
        
        Serializer *serializer = [Serializer new];
        NSArray *decodedObjects = [serializer serializeJSONObject:result];
        comp(decodedObjects);
        
    }];
    
    return 0;
}

@end

File: Demos/Todos/Swift/Todos/Fetcher.swift
//
//  Fetcher.swift
//  Todos
//
//  Created by Sasa Prodribaba on 25/12/2018.
//  Copyright © 2018 Sasa Prodribaba. All rights reserved.
//

import Foundation

class Fetcher {
    
    func fetchFromURL(url: URL, completion: @escaping (_ result: Data?) -> Void) {
        let session = URLSession.shared
        let dataTask = session.dataTask(with: url, completionHandler: { data, response, error in
            if let error = error {
                print("Error: \(error.localizedDescription)")
            } else {
                completion(data)
            }
        })
        dataTask.resume()
    }
    
    func fetchWithResult(completion: @escaping (_ result: [Todo]) -> Void) {
        guard let url = URL(string: Constants.todosEndpoint) else { return }
        self.fetchFromURL(url: url) {
            result in
            let serializer = Serializer()
            if let result = result {
                let decodedObjects = serializer.serializeJSONObject(jsonData: result)
                completion(decodedObjects)
            }
        }
    }
    
}
